### PR TITLE
Render a nicer 404

### LINKED ISSUE #88: Make the 404 page smart, e.g. search, "Did you mean"
Our not found page is currently static, but typically we can infer things from the URL that allow us to do more than just 'block' a user, e.g.

- break the URL up into parts and do a search: `pantsbuild.org/something-here` could turn into a search for `something here`, as would `pantsbuild.org/docs/v2.19/something-here` (i.e. ignore obvious 'cruft' like `docs`, `reference` or version slugs)
- bonus, iterating on #10: if it contains `/v1.`, they're likely looking for v1 docs so direct to v1.pantsbuild.org (potentially including a `site:https://v1.pantsbuild.org` search) (https://github.com/pantsbuild/pantsbuild.org/pull/87#issuecomment-1888007661)

Example: https://doc.rust-lang.org/stable/std/tree

### PR DESCRIPTION
**Pull request by thejcannon**
Fixes #88 by swizzling `NotFound/Content`

The important thing to me was catching `/v1` URLs

### CODE REVIEW

File: src/theme/NotFound/Content/index.jsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+          <Heading as="h1" className="hero__title">
+            Page Not Found
+          </Heading>
+          <p>We could not find what you were looking for in any of our pockets.</p>
+
+
+          {location.pathname.startsWith("/v1") ?
+            <p>The V1 Pants docs can be found at <Link to="https://v1.pantsbuild.org/">https://v1.pantsbuild.org/</Link></p> :
+            <span>
+              <p>Our excellent <Link to="/search">search page</Link> might be helpful.</p>
```
Comment (huonw) [inline src/theme/NotFound/Content/index.jsx:22]: Happy to do this as a future enhancement, but I think we can make this even better by pre-filling the search if we can.

```javascript
// a component of a URL that might be interesting is any sequence of things separated by these punctuations
const componentsRe = /[^-/ ]+/g;
// several things that appear in URLs are not interesting, e.g. /docs/foo should become just "foo", not "docs foo", similarly "pants" generally won't be informative, nor will a version specifier
const boringComponents = /^(doc|docs|reference|pants|v?[0-9]+\.[x0-9]+)$/

const estimateSearch = (location) => {
    // This site generally use the path and hash/fragment, so look at them to find any tidbits that may suggest what the user was looking for
    const strings = [location.pathname, location.hash];
    const allComponents = strings.map(decodeURI).flatMap(s => [...s.matchAll(componentsRe)].flat());
    const interestingComponents = allComponents.filter(s => !s.match(boringComponents))
    return interestingComponents.join(" ")
};
```

```suggestion
              <p>Try a <Link to={`/search?q=${encodeURIComponent(estimateSearch(location))}`}>search for {estimateSearch(location)}</Link>?</p>
```

**Review by huonw — COMMENTED** [review]
I like it!

File: src/theme/NotFound/Content/index.jsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    <main className={clsx("container margin-vert--xl", className)}>
+      <div className="row">
+        <div className="col col--6 col--offset-3">
+          <Heading as="h1" className="hero__title">
+            Page Not Found
+          </Heading>
+          <p>We could not find what you were looking for in any of our pockets.</p>
+
+
+          {location.pathname.startsWith("/v1") ?
```
[the remark this answers is not in the document]
Comment (thejcannon) [reply src/theme/NotFound/Content/index.jsx:19]: No, see https://docusaurus.io/docs/advanced/routing#generating-and-accessing-routes

(Also I tested this locally)

File: src/theme/NotFound/Content/index.jsx (hunk 1 above)
Comment (thejcannon) [reply src/theme/NotFound/Content/index.jsx:22]: Works like a charm

File: src/theme/NotFound/Content/index.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (huonw) [reply src/theme/NotFound/Content/index.jsx]: Hm, I _think_ that's something a bit different: using `useLocation()` ensures one has the page's current path when rendering server-side, whereas `window.location` might not even exist during SSR. It is still potentially decided statically and hard-coded into the output...

But, from running a build and inspecting 404.html, it looks like the 404 page _always_ runs client side (`404.html` doesn't have any of this content hard-coded in it, just the React set-up), so no `<BrowserOnly>` is necessary in this specific case.

[the remark this answers is not in the document]
Comment (thejcannon) [reply src/theme/NotFound/Content/index.jsx]: Huh, well TIL

[the remark this answers is not in the document]
Comment (huonw) [reply src/theme/NotFound/Content/index.jsx]: I tried this locally and I found there was a missing import of `BrowserOnly` (in addition to the formatting failure CI is highlighting)... also, it doesn't seem to work as intended, as the `<BrowserOnly>` content isn't rendered:

image

[the remark this answers is not in the document]
Comment (thejcannon) [reply src/theme/NotFound/Content/index.jsx]: Huh oddly I couldn't get
```
          <BrowserOnly>
            {() => {
              <span>Hello</span>
            }}
          </BrowserOnly>
```

to work (with import) from either `npm start` or `npm run build && npm run serve`.

Something must be very wrong.

[the remark this answers is not in the document]
Comment (thejcannon) [reply src/theme/NotFound/Content/index.jsx]: > Something must be very wrong.

I was missing a `return` :facepalm:

[the remark this answers is not in the document]
Comment (thejcannon) [reply src/theme/NotFound/Content/index.jsx]: Ok added the missing return and tested using `npm run build && npm run serve`

**Review by huonw — APPROVED** [review]
Whew, nice! Thanks for debugging the `return` issue.

**Comment by thejcannon** [comment]
It's one of those facepalm moments for sure.

Thanks to you for suggesting the code for the search!

**Pull request merged**